// hedron/lib/components/Container.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _templateObject = _taggedTemplateLiteral(["\n  display: flex;\n  ", "\n  ", "\n  ", "\n  ", "\n  ", "\n  ", "\n"], ["\n  display: flex;\n  ", "\n  ", "\n  ", "\n  ", "\n  ", "\n  ", "\n"]);

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _styledComponents = require("styled-components");

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _LayoutProvider = require("./LayoutProvider");

var _flex = require("../utils/flex");

var _utils = require("../utils");

var _passOn = require("../utils/passOn");

var _passOn2 = _interopRequireDefault(_passOn);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var containerDirectionError = "Container direction must be `horizontal` or `vertical`";

var directionToFlex = function directionToFlex(direction) {
  if (direction === "vertical") {
    return "column";
  } else if (direction === "horizontal") {
    return "row";
  }
  throw Error(containerDirectionError);
};

var compute = function compute(name) {
  return (0, _utils.breakpoint)(name, function (props, name) {
    return [(0, _utils.applyBreakpointStyle)('width', props.width, name), (0, _utils.applyBreakpointStyle)('height', props.height, name), (0, _utils.applyBreakpointStyle)('flex-grow', props.grow, name, { on: 1, off: 0 }), (0, _utils.applyBreakpointStyle)('flex-shrink', props.shrink, name, { on: 1, off: 0 }), (0, _utils.applyBreakpointStyle)('flex', props.flex, name), (0, _utils.applyBreakpointStyle)('flex-wrap', props.wrap, name, { on: 'wrap', off: 'nowrap' }), (0, _utils.applyBreakpointStyle)('flex-direction', props.direction, name, { modifier: function modifier(v) {
        return directionToFlex(v);
      } })];
  });
};

var Wrapper = _styledComponents2.default.div(_templateObject, function (props) {
  return props.alignX && (0, _flex.horizontalAlign)(props.alignX, props.direction);
}, function (props) {
  return props.alignY && (0, _flex.verticalAlign)(props.alignY, props.direction);
}, compute("xs"), compute("sm"), compute("md"), compute("lg"));

Wrapper.defaultProps = {
  direction: "horizontal"
};

var Container = function Container(_ref) {
  var children = _ref.children,
      props = _objectWithoutProperties(_ref, ["children"]);

  return _react2.default.createElement(
    Wrapper,
    props,
    (0, _passOn2.default)(children, function (child) {
      return {
        xs: child.props.xs || child.props.direction === "horizontal" && 12 / children.length
      };
    })
  );
};

exports.default = (0, _LayoutProvider.withLayout)(Container);
